src/Gui/ToolBarManager.cpp

```diff
@@ -294,10 +304,137 @@ class ToolBarArea : public QWidget
     boost::signals2::scoped_connection &_conn;
 };
 
+class ToolBar: public QToolBar
+{
+public:
+    void initStyleOption(QStyleOptionToolBar *option) const
+    {
+        QToolBar::initStyleOption(option);
+    }
+};
+
 } // namespace Gui
 
 // -----------------------------------------------------------
 
+ToolBarGrip::ToolBarGrip(QToolBar * parent)
+    :QWidget(parent)
+{
+    QStyle *style = parent->style();
+    QStyleOptionToolBar opt;
+    static_cast<ToolBar*>(parent)->initStyleOption(&opt);
+    opt.features = QStyleOptionToolBar::Movable;
+    int width = style->subElementRect(QStyle::SE_ToolBarHandle, &opt, parent).width();
+    this->setFixedWidth(width+4);
+
+    setCursor(Qt::OpenHandCursor);
+    setMouseTracking(true);
+    auto actions = parent->actions();
+    _action = parent->insertWidget(actions.isEmpty()?nullptr:actions[0], this);
+    setVisible(true);
+}
+
+QAction *ToolBarGrip::action() const
+{
+    return _action;
+}
+
+void ToolBarGrip::paintEvent(QPaintEvent*)
+{
+    QPainter painter(this);
+    if (auto toolbar = qobject_cast<QToolBar*>(parentWidget())) {
+        QStyle *style = toolbar->style();
+        QStyleOptionToolBar opt;
+        static_cast<ToolBar*>(toolbar)->initStyleOption(&opt);
+        opt.features = QStyleOptionToolBar::Movable;
+        // opt.rect = style->subElementRect(QStyle::SE_ToolBarHandle, &opt, toolbar);
+        opt.rect = this->rect();
+        style->drawPrimitive(QStyle::PE_IndicatorToolBarHandle, &opt, &painter, toolbar);
+    }
+    else {
+        painter.setPen(Qt::transparent);
+        painter.setOpacity(0.5);
+        painter.setBrush(QBrush(Qt::black, Qt::Dense6Pattern));
+        QRect rect(this->rect());
+        painter.drawRect(rect);
+    }
+}
+
+void ToolBarGrip::mouseMoveEvent(QMouseEvent *me)
+{
+    auto toolbar = qobject_cast<QToolBar*>(parentWidget());
+    if (!toolbar) {
+        return;
+    }
+    auto area = ToolBarManager::getInstance()->getToolBarArea(toolbar);
+    if (!area)
+        return;
+
+    QPoint pos = me->globalPos();
+    QRect rect(toolbar->mapToGlobal(QPoint(0,0)), toolbar->size());
+    if (rect.contains(pos))
+        return;
+
+    // If mouse is out of the toolbar, remove the toolbar from area and float the
+    // toolbar
+    {
+        QSignalBlocker blocker(toolbar);
+        area->removeWidget(toolbar);
+        getMainWindow()->addToolBar(toolbar);
+        toolbar->setWindowFlags(Qt::Tool
+                | Qt::FramelessWindowHint
+                | Qt::X11BypassWindowManagerHint);
+        toolbar->adjustSize();
+        pos = toolbar->mapFromGlobal(pos);
+        toolbar->move(pos.x()-10, pos.y()-10);
+        toolbar->setVisible(true);
+    }
+    toolbar->topLevelChanged(true);

```

Comment:
While you don't have to mark signals with explicit `Q_EMIT` it is nice at it clearly states the intention of that call. Also a comment why it is outside of the block above (to not block this signal - a bug that you have in the previous PR) would be nice as it once again is hard to quickly understand why every call is in that block and it is outside.